Complete sprint 12

Evo-Eclipse opened the pull request:
Domain Functionality:
- Implemented the main feed feature with real images from Unsplash API
- Added like/unlike functionality for photos with server synchronization
- Implemented full-size image viewing with proper zooming and sharing options
- Added pagination with infinite scrolling for seamless user experience

Network Layer Enhancements:
- Added `ImagesListService` for fetching photos from Unsplash API with pagination support
- Implemented adaptive pagination that automatically adjusts `itemsPerPage` based on network performance
- Added like functionality with proper error handling and UI feedback
- Enhanced network debugging with improved logging across all service classes

Data Model & Storage:
- Created new `Photo` model with size, URLs, and metadata support
- Added `Equatable` conformance to `Photo` for easier comparison
- Implemented proper date formatting with ISO8601DateFormatter
- Added tracking of network request durations for performance optimization

UI Improvements:
- Refactored `ImagesListViewController` to use real API data instead of local images
- Enhanced `SingleImageViewController` with proper image scaling, zooming, and rotation support
- Added placeholder images for the feed during loading
- Implemented image caching with Kingfisher across the app
- Added error alert handling with retry options for failed image loads

User Experience:
- Added loading indicators during network operations
- Implemented infinite scrolling with automatic pagination
- Added logout functionality with proper cleanup of data, cookies, and caches
- Added confirmation dialogs for important user actions

Code Quality:
- Removed unnecessary whitespace and empty lines throughout codebase
- Improved error handling with explicit error types and descriptive messages

Performance Optimizations:
- Implemented image preloading with Kingfisher
- Added cancellation of in-progress downloads when cells are recycled
- Optimized table view updates with animated batch operations
- Added image transition animations for smoother user experience

Testing & Debugging:
- Added debug print statements for network operations
- Improved error recovery flows with user-friendly alerts
- Added progressive loading with performance metrics collection

Evo-Eclipse committed:
Implement Image List flow

- Added ImagesListService to fetch photos from Unsplash
- Implemented pagination with adaptive request sizing
- Created Photo model for storing image data and metadata
- Modified ImagesListViewController to use actual photos instead of static files
- Added Kingfisher integration for image loading and caching
- Added placeholder for images during loading

Evo-Eclipse committed:
Implement Like Photo feature

- Added like functionality to images in list
- Created ImagesListCellDelegate for handling like interactions
- Implemented like/unlike network request in ImagesListService
- Added notification system for like status updates
- Updated UI to reflect like status changes dynamically

Evo-Eclipse committed:
Remove optimistic like button UI updates

- Removed premature UI state changes for like functionality
- Ensures more accurate representation of like status
- Prevents UI from updating before backend confirmation

Evo-Eclipse committed:
Refactor Like Functionality & Improve Error Handling

- Replaced notification-based like status updates with completion handler
- Added UIBlockingProgressHUD to show loading state during like operations
- Introduced dedicated error alert for like status change failures
- Removed static notification names related to like status changes

Evo-Eclipse committed:
Enhance Single Image View & Implement Full Image Loading

- Added support for loading high-quality images from full URL
- Implemented progressive image loading with loading indicator
- Enhanced image scaling and centering logic in scroll view
- Added error handling with retry option for image loading
- Updated ImagesListViewController to pass full image URL
- Optimized image view scaling and content positioning

Evo-Eclipse committed:
Implement Logout flow

- Created ProfileLogoutService to centralize logout functionality
- Implemented thorough cleanup of user-related data and web cookies
- Added logout confirmation alert in ProfileViewController
- Refined alert in SplashViewController and SingleImageViewController
- Replaced fatalError with assertionFailure in critical sections

Evo-Eclipse committed:
Minor Swift syntax improvements

- Remove redundant `adjustImageScale` call in SingleImageViewController
- Minor constraint and syntax cleanups in various view controllers

NaR0RoW commented on ImageFeed/Application/AppDelegate.swift:
```diff
@@ -25,7 +23,4 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         // If any sessions were discarded while the application was not running, this will be called shortly after application:didFinishLaunchingWithOptions.
```
Можно удалить не используемые методы

NaR0RoW commented on ImageFeed/Application/AppDelegate.swift:
```diff
@@ -3,8 +3,6 @@ import UIKit
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
```
Стоит отметить как `final`. Если класс не наследуется, то лучше отметить его как `final` для ускорения работы компилятора

NaR0RoW commented on .gitignore:
```diff
@@ -4,4 +4,4 @@
 DerivedData/
 .swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
-**/Constants.swift
+# **/Constants.swift
```
Можно тогда удалить эту строку

NaR0RoW commented on ImageFeed/Application/SceneDelegate.swift:
```diff
@@ -43,7 +42,4 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         // Use this method to save data, release shared resources, and store enough scene-specific state information
```
В этом файле тоже можно удалить все не используемые методы и поставить `final` у `SceneDelegate`

NaR0RoW commented on ImageFeed/Controllers/ImagesListViewController.swift:
```diff
@@ -75,34 +142,41 @@ final class ImagesListViewController: UIViewController {
 extension ImagesListViewController: UITableViewDataSource {
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return photosName.count
+        return photos.count
```
В однострочных выражениях `return` лучше опускать

NaR0RoW commented on ImageFeed/Resources/Assets.xcassets/Images/placeholder.feed.imageset/placeholder_feed.png:
Картинки 1x лучше не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone 4. Он не поддерживает минимальную версию приложения, поэтому можно смело удалять, чтобы не занимать лишнюю память

NaR0RoW commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,287 @@
+        }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
```
Можно создать отдельный `enum` типа `String` для http методов

Evo-Eclipse replied:
Creating a String enum for HTTP methods here would violate KISS, YAGNI principles. Since these methods are only used in a single location, introducing an enum adds unnecessary complexity without any clear benefits.

NaR0RoW replied:
Ты прав, что следует придерживаться `KISS` и `YAGNI`, но чуть дальше, в методе `makeLikeRequest` есть следующий код:
``` swift
request.httpMethod = isLike ? "POST" : "DELETE"
``` 
Т.е. как минимум `enum` будет использоваться в трех местах, и, спойлер, в будущих спринтах их будет больше. В данном случае `enum` - это не усложнение, а стандарт, который уменьшает количество багов и делает код более читаемым. На реальных проектах практически всегда используют такие enum'ы. Например библиотеки Alamofire, Moya, даже в документации от Apple и open-source проектах - это базовая практика. Поэтому в данном случае enum - это хорошее решение и часть архитектурной зрелости.

Из своего опыта и из общения с множеством коллег могу сказать, что не всегда стоит слишком буквально и жёстко придерживаться принципов вроде KISS, YAGNI, SOLID, ООП, архитектур и т.д. На бумаге всё выглядит логично и стройно, но в реальности всё куда сложнее.
Часто именно попытки строго следовать этим принципам делают код чрезмерно абстрактным, сложным для понимания и сопровождения. У каждого разработчика своё понимание этих принципов, и оно не всегда совпадает с реальной задачей или контекстом.
Ты можешь прийти в условный ВК (или любой крупный проект) и увидеть файлы по несколько тысяч строк, где вроде бы одновременно всё и соблюдено, и нарушено. И тебе придётся потратить несколько дней или даже недель, чтобы просто разобраться в происходящем.
Поэтому это в каком-то смысле холиварный вопрос. Очень хорошо, что ты сейчас стараешься писать чисто и думаешь о принципах - это правильно. Но важно также помнить, что их не стоит возводить в абсолют. Контекст, читаемость и здравый смысл часто важнее теоретической чистоты.

Evo-Eclipse committed:
Improve date handling & Refactor Constants

- Made Photo.createdAt optional to safely handle nil dates
- Enhanced date formatting logic in ImagesListViewController
- Changed var secretKey to let in Constants.swift.example
- Removed force unwrap for defaultBaseURL in Constants
- Added final keyword to AppDelegate and SceneDelegate
- Simplified return statement in numberOfRowsInSection
